File: material/src/main/scala/material/plugin/config/S3PackageMaterialConfiguration.scala
package material.plugin.config

import com.thoughtworks.go.plugin.api.material.packagerepository._
import com.thoughtworks.go.plugin.api.response.Result
import com.thoughtworks.go.plugin.api.response.validation.{ValidationError, ValidationResult}
import com.thoughtworks.go.plugin.api.config.Property

import scala.collection.JavaConverters._
import org.apache.commons.lang3.StringUtils
import material.plugin.S3PackageMaterialPoller

case class Config(name: String, displayName: String, order: Int) {
  def toPackageProperty = new PackageMaterialProperty(name)
    .`with`[String](Property.DISPLAY_NAME, displayName)
    .`with`[Integer](Property.DISPLAY_ORDER, order)

  def toProperty = new Property(name)
    .`with`[String](Property.DISPLAY_NAME, displayName)
    .`with`[Integer](Property.DISPLAY_ORDER, order)
}



object S3PackageMaterialConfiguration {
  val S3_BUCKET = "S3_BUCKET"
  val S3_ACCESS_KEY_ID = "S3_AWS_ACCESS_KEY_ID"
  val S3_SECRET_ACCESS_KEY = "S3_AWS_SECRET_ACCESS_KEY"
  val ARTIFACT_NAME = "ARTIFACT_NAME"
  val repoConfigs = List(
    Config(S3_BUCKET, "S3 Bucket", 0),
    Config(S3_ACCESS_KEY_ID, "S3 Access Key ID", 1),
    Config(S3_SECRET_ACCESS_KEY, "S3 Secret Access Key", 2)
  )
  
  val packageConfigs = List(
    Config(ARTIFACT_NAME, "Artifact Name", 0)
  )
}

class S3PackageMaterialConfiguration extends PackageMaterialConfiguration {
  override def getRepositoryConfiguration: RepositoryConfiguration = {
    val repoConfig = new RepositoryConfiguration()
    S3PackageMaterialConfiguration.repoConfigs.map(_.toPackageProperty).foreach(p => repoConfig.add(p))
    repoConfig
  }

  def validate(repoConfig: RepositoryConfiguration, property: String, message: String) = {
    if(repoConfig.get(property) == null || StringUtils.isBlank(repoConfig.get(property).getValue)) {
      val validationResult = new ValidationResult()
      validationResult.addError(new ValidationError(property, message))
      Some(validationResult)
    }else{
      None
    }
  }

  override def isRepositoryConfigurationValid(repoConfig: RepositoryConfiguration): ValidationResult = {
    val errors = List(validate(repoConfig, S3PackageMaterialConfiguration.S3_BUCKET, s"${S3PackageMaterialConfiguration.S3_BUCKET} configuration is missing or value is empty"),
      validate(repoConfig, S3PackageMaterialConfiguration.S3_ACCESS_KEY_ID, s"${S3PackageMaterialConfiguration.S3_ACCESS_KEY_ID} configuration is missing or value is empty"),
      validate(repoConfig, S3PackageMaterialConfiguration.S3_SECRET_ACCESS_KEY, s"${S3PackageMaterialConfiguration.S3_SECRET_ACCESS_KEY} configuration is missing or value is empty")
    ).flatMap(vr => vr.map(_.getErrors.asScala).getOrElse(List[ValidationError]()))

    val validationResult = new ValidationResult()
    validationResult.addErrors(errors.asJava)
    validationResult
  }

  override def isPackageConfigurationValid(packageConfig: PackageConfiguration, repoConfig: RepositoryConfiguration): ValidationResult = {
    val errors = List(
      validate(repoConfig, S3PackageMaterialConfiguration.ARTIFACT_NAME, s"${S3PackageMaterialConfiguration.ARTIFACT_NAME} configuration is missing or value is empty")
    ).flatMap(vr => vr.map(_.getErrors.asScala).getOrElse(List[ValidationError]()))
    val validationResult = new ValidationResult()
    validationResult.addErrors(errors.asJava)
    validationResult
  }

  override def getPackageConfiguration: PackageConfiguration = {
    val packageConfig = new PackageConfiguration()
    S3PackageMaterialConfiguration.packageConfigs.map(_.toPackageProperty).foreach(p => packageConfig.add(p))
    packageConfig
  }
}

File: material/src/main/java/com/indix/gocd/s3material/config/S3PackageMaterialConfiguration.java
package com.indix.gocd.s3material.config;

import com.indix.gocd.utils.utils.Function;
import com.indix.gocd.utils.utils.Functions;
import com.indix.gocd.utils.utils.Lists;
import com.thoughtworks.go.plugin.api.config.Configuration;
import com.thoughtworks.go.plugin.api.config.Property;
import com.thoughtworks.go.plugin.api.logging.Logger;
import com.thoughtworks.go.plugin.api.material.packagerepository.PackageConfiguration;
import com.thoughtworks.go.plugin.api.material.packagerepository.PackageMaterialConfiguration;
import com.thoughtworks.go.plugin.api.material.packagerepository.RepositoryConfiguration;
import com.thoughtworks.go.plugin.api.response.validation.ValidationError;
import com.thoughtworks.go.plugin.api.response.validation.ValidationResult;
import org.apache.commons.lang3.StringUtils;

import java.util.ArrayList;
import java.util.List;

public class S3PackageMaterialConfiguration implements PackageMaterialConfiguration {
    public static String S3_BUCKET = "S3_BUCKET";
    public static String PIPELINE_NAME = "PIPELINE_NAME";
    public static String STAGE_NAME = "STAGE_NAME";
    public static String JOB_NAME = "JOB_NAME";
    private Logger LOG = Logger.getLoggerFor(this.getClass());

    public static List<Config> repoConfigs = new ArrayList<Config>() {{
        add(new Config(S3_BUCKET, "S3 Bucket", 0));
    }};

    public static List<Config> packageConfigs = new ArrayList<Config>(){{
        add(new Config(PIPELINE_NAME, "Pipeline Name", 0));
        add(new Config(STAGE_NAME, "Stage Name", 1));
        add(new Config(JOB_NAME, "Job Name", 2));
    }};


    Function<Config, Property> toPackageProperty = new Function<Config, Property>() {
        @Override
        public Property apply(Config input) {
            return input.toPackageProperty();
        }
    };

    @Override
    public RepositoryConfiguration getRepositoryConfiguration() {
        final RepositoryConfiguration repoConfig = new RepositoryConfiguration();
        List<Property> properties = Lists.map(S3PackageMaterialConfiguration.repoConfigs, toPackageProperty);
        Lists.foreach(properties, new Functions.VoidFunction<Property>() {
            @Override
            public void execute(Property property) {
                repoConfig.add(property);
            }
        });
        return repoConfig;
    }

    @Override
    public PackageConfiguration getPackageConfiguration() {
        final PackageConfiguration packageConfig = new PackageConfiguration();
        List<Property> properties = Lists.map(S3PackageMaterialConfiguration.packageConfigs, toPackageProperty);
        Lists.foreach(properties, new Functions.VoidFunction<Property>() {
            @Override
            public void execute(Property property) {
                packageConfig.add(property);
            }
        });
        return packageConfig;
    }

    @Override
    public ValidationResult isRepositoryConfigurationValid(final RepositoryConfiguration repositoryConfiguration) {
        return runValidations(repositoryConfiguration, repoConfigs);
    }

    @Override
    public ValidationResult isPackageConfigurationValid(final PackageConfiguration packageConfiguration, RepositoryConfiguration repositoryConfiguration) {
        return runValidations(packageConfiguration, packageConfigs);
    }

    private static ValidationResult runValidations(final Configuration config, List<Config> configsToBeCheckedFor) {
        List<ValidationResult> validationResults = Lists.map(configsToBeCheckedFor, new Function<Config, ValidationResult>() {
            @Override
            public ValidationResult apply(Config input) {
                return validate(config,
                        input.getName(),
                        String.format("%s configuration is missing or value is empty", input.getName()),
                        input.isRequired());
            }
        });
        List<ValidationError> validationErrors = Lists.flatMap(validationResults, new Function<ValidationResult, List<ValidationError>>() {
            @Override
            public List<ValidationError> apply(ValidationResult input) {
                return input.getErrors();
            }
        });
        ValidationResult finalResult = new ValidationResult();
        finalResult.addErrors(validationErrors);
        return finalResult;
    }

    private static ValidationResult validate(Configuration config, String property, String message, Boolean required) {
        if(required && (config.get(property) == null || StringUtils.isBlank(config.get(property).getValue()))) {
            ValidationResult validationResult = new ValidationResult();
            validationResult.addError(new ValidationError(property, message));
            return validationResult;
        }else{
            return new ValidationResult();
        }
    }
}
